### PR TITLE
Add R2 score to metrics

### LINKED ISSUE #8085: Add Average Precision and R2 to the metrics
### Discussed in https://github.com/Project-MONAI/MONAI/discussions/8081

<div type='discussions-op-text'>

<sup>Originally posted by **thibaultdvx** September 13, 2024</sup>
Hi, 
I would like to suggest to add the [**Average Precision**](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html) in the classification metrics. I think that the implementation would be very similar to ROC AUC, so it wouldn't be a big deal to integrate it.
I would be happy to make a PR to do so.

I also feel that an important regression metric is missing:[ **R2 score**](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html). Is there any reason why?
Once again, I would be happy to help to implement it!

WDYT?</div>

### PR DESCRIPTION
**Pull request by thibaultdvx**
Fixes #8085.

### Description

I created:
- `R2Metric` and `compute_r2_score` in `monai.metrics`,
- a handler called `R2Score` in `monai.handlers`,
- three unittest modules: `test_compute_r2_score.py`, `test_handler_r2_score.py` and `test_handler_r2_score_dist.py`.

I also modified the docs to mention R2 score.

Note:  
- R2 score expects 1D or 2D data, i.e. I didn't allow the computation of R2 at pixel level in an image, like it can be done with the other regression metrics. 
- One needs all the data to compute R2 score (because of the mean), so tensors are simply accumulated and the metric computation is made once on all the data.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by thibaultdvx** [comment]
hi @KumoLiu, could you please have a look at this PR?

File: tests/test_handler_r2_score.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import unittest
+
+import numpy as np
+import torch
+
+from monai.handlers import R2Score
+
+
+class TestHandlerR2Score(unittest.TestCase):
```
Comment (KumoLiu) [inline tests/test_handler_r2_score.py:22]: Base on the ci :https://github.com/Project-MONAI/MONAI/actions/runs/10926409391/job/30330218578?pr=8093#step:8:13844

You need skip the tests if no ignite installed.

File: monai/metrics/r2_score.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    if not isinstance(p, int) or p < 0:
+        raise ValueError(f"`p` must be an integer larger or equal to 0, got {p}.")
+
+    return multi_output, p
+
+
+def _calculate(y_pred: np.ndarray, y: np.ndarray, p: int) -> float:
+    num_obs = len(y)
+    rss = np.sum((y_pred - y) ** 2)
+    tss = np.sum(y**2) - np.sum(y) ** 2 / num_obs
```
Comment (KumoLiu) [inline monai/metrics/r2_score.py:114]: Should be `tss = np.sum((y - np.mean(y)) ** 2)` here?

**Review by KumoLiu — COMMENTED** [review]
Thanks for the PR, leave few comments inline.

File: monai/handlers/r2_score.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+from collections.abc import Callable
+
+from monai.handlers.ignite_metric import IgniteMetricHandler
+from monai.metrics import R2Metric
+from monai.utils import MultiOutput
+
+
+class R2Score(IgniteMetricHandler):
+    """
+    Computes :math:`R^{2}` score accumulating predictions and the ground-truth during an epoch and applying `compute_r2_score`.
```
Comment (ericspod) [inline monai/handlers/r2_score.py:23]: As elsewhere it would help to explain where this is used and what it would be for. The math is explained in the function which is fine so no need here, just a high level idea.

File: monai/metrics/r2_score.py (hunk 2 above)
Comment (thibaultdvx) [reply monai/metrics/r2_score.py]: Both formula works, but I agree that yours is clearer! I made the change

File: tests/test_handler_r2_score_dist.py
[the remark this answers is not in the document]
Comment (thibaultdvx) [reply tests/test_handler_r2_score_dist.py:1]: Done!

File: monai/handlers/r2_score.py (hunk 3 above)
Comment (thibaultdvx) [reply monai/handlers/r2_score.py:23]: Hi @ericspod, tell me if the new version looks good to you!

File: tests/test_handler_r2_score.py (hunk 1 above)
Comment (thibaultdvx) [reply tests/test_handler_r2_score.py]: Done, I changed `min_tests.py`.

**Comment by thibaultdvx** [comment]
@KumoLiu @ericspod ready for review

**Review by ericspod — APPROVED** [review]
I think this is good to go.

File: monai/handlers/r2_score.py (hunk 3 above)
Comment (ericspod) [reply monai/handlers/r2_score.py:23]: I mean more to explain what the R2 means, what it would be used with (eg. compare classifications vs. segmentations), what it's value is vs other metrics. I would put this in the docstring for `R2Metric` however. Other than that this is fine here.
